File: jira/create_jira_ticket.py
#!/usr/bin/env python3
"""
Create JIRA tickets using static configuration from jira_fields_config.json

Usage:
    python create_jira_ticket.py --summary "Fix login bug" --description "Users can't log in"
    python create_jira_ticket.py --summary "Add new feature" --description "..." --dry-run
"""

import argparse
import json
import os
import sys
from typing import Dict, Any, Optional
import requests
from requests.auth import HTTPBasicAuth
from dotenv import load_dotenv

# Load environment variables from .env file (look in parent directory too)
load_dotenv()
load_dotenv(os.path.join(os.path.dirname(__file__), '..', '.env'))


def load_config(config_path: str = "jira_fields_config.json") -> Dict[str, Any]:
    """Load field configuration from JSON file."""
    if not os.path.exists(config_path):
        print(f"Error: Config file not found at {config_path}")
        sys.exit(1)

    with open(config_path, 'r') as f:
        return json.load(f)


def get_jira_credentials() -> tuple[str, str, str]:
    """Get JIRA credentials from environment variables."""
    jira_url = os.getenv('JIRA_URL')
    jira_username = os.getenv('JIRA_USERNAME')
    jira_api_token = os.getenv('JIRA_API_TOKEN')

    if not all([jira_url, jira_username, jira_api_token]):
        print("Error: Missing JIRA credentials in environment variables")
        print("Required: JIRA_URL, JIRA_USERNAME, JIRA_API_TOKEN")
        sys.exit(1)

    return jira_url, jira_username, jira_api_token


def create_jira_ticket(
    summary: str,
    description: str,
    dry_run: bool = False,
    config_path: str = "jira_fields_config.json"
) -> Optional[Dict[str, Any]]:
    """
    Create a JIRA ticket using fields from config file.

    Args:
        summary: Ticket summary/title
        description: Ticket description
        dry_run: If True, print payload without creating ticket
        config_path: Path to config file

    Returns:
        Created issue data or None if dry run or error
    """
    # Load configuration
    config = load_config(config_path)

    # Get fields from config
    components = config.get('components', [])
    labels = config.get('labels', [])
    team = config.get('team', '')
    team_field = config.get('team_field')  # Custom field ID for team

    # Get JIRA credentials
    jira_url, jira_username, jira_api_token = get_jira_credentials()
    auth = HTTPBasicAuth(jira_username, jira_api_token)

    # Build JIRA payload
    payload = {
        "fields": {
            "project": {"key": "RHCLOUD"},
            "summary": summary,
            "description": description,
            "issuetype": {"name": "Story"},
            "labels": labels,
        }
    }

    # Add components if any
    if components:
        payload["fields"]["components"] = [{"name": comp} for comp in components]

    # Add team if field ID is configured
    # Try different formats - some JIRA instances want just the string, others want an ID
    if team and team_field:
        # If team looks like an ID (starts with digits), use it as-is
        # Otherwise, try as a string value (for text fields)
        if team.isdigit() or team.startswith('team-'):
            payload["fields"][team_field] = team
        else:
            # For text/string fields, just use the string directly
            payload["fields"][team_field] = team

    if dry_run:
        print("\n" + "="*80)
        print("DRY RUN - Would create ticket with the following payload:")
        print("="*80)
        print(json.dumps(payload, indent=2))
        print("="*80)
        print(f"\nTeam: {team}")
        print(f"Labels: {', '.join(labels)}")
        print(f"Components: {', '.join(components)}")
        print(f"JIRA URL: {jira_url}")
        print(f"Username: {jira_username}")
        return None

    # Create the ticket
    try:
        response = requests.post(
            f"{jira_url}/rest/api/2/issue",
            auth=auth,
            headers={'Content-Type': 'application/json'},
            json=payload
        )
        response.raise_for_status()

        result = response.json()
        issue_key = result.get('key')

        print(f"\n✓ Successfully created ticket: {issue_key}")
        print(f"  URL: {jira_url}/browse/{issue_key}")
        print(f"  Team: {team}")
        print(f"  Labels: {', '.join(labels)}")
        print(f"  Components: {', '.join(components)}")

        return result

    except requests.exceptions.RequestException as e:
        print(f"\n✗ Error creating ticket: {e}")
        if hasattr(e, 'response') and e.response is not None:
            try:
                error_detail = e.response.json()
                print(f"  Details: {json.dumps(error_detail, indent=2)}")
            except:
                print(f"  Response: {e.response.text}")
        return None


def main():
    parser = argparse.ArgumentParser(
        description='Create JIRA tickets using configuration from jira_fields_config.json',
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog="""
Examples:
  # Create a ticket with dry run
  python create_jira_ticket.py \\
      --summary "Fix login bug" \\
      --description "Users cannot log in after upgrade" \\
      --dry-run

  # Create a real ticket
  python create_jira_ticket.py \\
      --summary "Add new dashboard" \\
      --description "Create analytics dashboard with real-time metrics"

Note: Components, labels, and team are read from jira_fields_config.json
        """
    )

    parser.add_argument('--summary', required=True, help='Ticket summary/title')
    parser.add_argument('--description', required=True, help='Ticket description')
    parser.add_argument('--dry-run', action='store_true', help='Print payload without creating ticket')
    parser.add_argument('--config', default='jira_fields_config.json', help='Path to config file')

    args = parser.parse_args()

    create_jira_ticket(
        summary=args.summary,
        description=args.description,
        dry_run=args.dry_run,
        config_path=args.config
    )


if __name__ == '__main__':
    main()

File: jira/create-jira-ticket.go
package main

import (
	"bytes"
	"encoding/json"
	"flag"
	"fmt"
	"io"
	"net/http"
	"os"
	"path/filepath"

	"github.com/joho/godotenv"
)

// Config represents the JIRA fields configuration
type Config struct {
	Components []string `json:"components"`
	Labels     []string `json:"labels"`
	TeamField  string   `json:"team_field"`
	Team       string   `json:"team"`
}

// JiraPayload represents the JIRA issue creation payload
type JiraPayload struct {
	Fields map[string]interface{} `json:"fields"`
}

// JiraResponse represents the response from JIRA
type JiraResponse struct {
	Key string `json:"key"`
}

func main() {
	// Command-line flags
	summary := flag.String("summary", "", "Ticket summary/title (required)")
	description := flag.String("description", "", "Ticket description (required)")
	dryRun := flag.Bool("dry-run", false, "Print payload without creating ticket")
	configPath := flag.String("config", "jira_fields_config.json", "Path to config file")
	postToSlackFlag := flag.Bool("post-to-slack", false, "Post ticket to Slack workflow")
	flag.Parse()

	if *summary == "" || *description == "" {
		fmt.Println("Error: --summary and --description are required")
		flag.Usage()
		os.Exit(1)
	}

	// Load .env from current directory and parent directory
	godotenv.Load()
	godotenv.Load(filepath.Join("..", ".env"))

	// Get JIRA credentials from environment
	jiraURL := os.Getenv("JIRA_URL")
	jiraUsername := os.Getenv("JIRA_USERNAME")
	jiraAPIToken := os.Getenv("JIRA_API_TOKEN")

	if jiraURL == "" || jiraUsername == "" || jiraAPIToken == "" {
		fmt.Println("Error: Missing JIRA credentials in environment variables")
		fmt.Println("Required: JIRA_URL, JIRA_USERNAME, JIRA_API_TOKEN")
		os.Exit(1)
	}

	// Load config
	config, err := loadConfig(*configPath)
	if err != nil {
		fmt.Printf("Error loading config: %v\n", err)
		os.Exit(1)
	}

	// Build JIRA payload
	payload := buildPayload(*summary, *description, config)

	if *dryRun {
		printDryRun(payload, config, jiraURL, jiraUsername)
		return
	}

	// Create the ticket
	issueKey, err := createTicket(jiraURL, jiraUsername, jiraAPIToken, payload)
	if err != nil {
		fmt.Printf("\n✗ Error creating ticket: %v\n", err)
		os.Exit(1)
	}

	// Print success
	fmt.Printf("\n✓ Successfully created ticket: %s\n", issueKey)
	ticketURL := fmt.Sprintf("%s/browse/%s", jiraURL, issueKey)
	fmt.Printf("  URL: %s\n", ticketURL)
	fmt.Printf("  Team: %s\n", config.Team)
	fmt.Printf("  Labels: %s\n", joinStrings(config.Labels))
	fmt.Printf("  Components: %s\n", joinStrings(config.Components))

	// Post to Slack if flag is set
	if *postToSlackFlag {
		slackWebhookURL := os.Getenv("SLACK_WEBHOOK_URL")
		if slackWebhookURL == "" {
			fmt.Printf("\n⚠ Warning: SLACK_WEBHOOK_URL not set in environment variables\n")
		} else {
			if err := postToSlack(slackWebhookURL, issueKey, ticketURL, *summary, *description, config); err != nil {
				fmt.Printf("\n⚠ Warning: Failed to post to Slack: %v\n", err)
			} else {
				fmt.Printf("\n✓ Posted to Slack\n")
			}
		}
	}
}

func loadConfig(path string) (*Config, error) {
	file, err := os.Open(path)
	if err != nil {
		return nil, fmt.Errorf("config file not found at %s", path)
	}
	defer file.Close()

	var config Config
	if err := json.NewDecoder(file).Decode(&config); err != nil {
		return nil, fmt.Errorf("failed to parse config: %v", err)
	}

	return &config, nil
}

func buildPayload(summary, description string, config *Config) *JiraPayload {
	fields := map[string]interface{}{
		"project":     map[string]string{"key": "RHCLOUD"},
		"summary":     summary,
		"description": description,
		"issuetype":   map[string]string{"name": "Story"},
		"labels":      config.Labels,
	}

	// Add components if any
	if len(config.Components) > 0 {
		components := make([]map[string]string, len(config.Components))
		for i, comp := range config.Components {
			components[i] = map[string]string{"name": comp}
		}
		fields["components"] = components
	}

	// Add team if field ID is configured
	if config.Team != "" && config.TeamField != "" {
		fields[config.TeamField] = config.Team
	}

	return &JiraPayload{Fields: fields}
}

func printDryRun(payload *JiraPayload, config *Config, jiraURL, username string) {
	fmt.Println("\n" + strings("=", 80))
	fmt.Println("DRY RUN - Would create ticket with the following payload:")
	fmt.Println(strings("=", 80))

	jsonBytes, _ := json.MarshalIndent(payload, "", "  ")
	fmt.Println(string(jsonBytes))

	fmt.Println(strings("=", 80))
	fmt.Printf("\nTeam: %s\n", config.Team)
	fmt.Printf("Labels: %s\n", joinStrings(config.Labels))
	fmt.Printf("Components: %s\n", joinStrings(config.Components))
	fmt.Printf("JIRA URL: %s\n", jiraURL)
	fmt.Printf("Username: %s\n", username)
}

func createTicket(jiraURL, username, apiToken string, payload *JiraPayload) (string, error) {
	url := fmt.Sprintf("%s/rest/api/2/issue", jiraURL)

	jsonData, err := json.Marshal(payload)
	if err != nil {
		return "", fmt.Errorf("failed to marshal payload: %v", err)
	}

	req, err := http.NewRequest("POST", url, bytes.NewBuffer(jsonData))
	if err != nil {
		return "", fmt.Errorf("failed to create request: %v", err)
	}

	req.SetBasicAuth(username, apiToken)
	req.Header.Set("Content-Type", "application/json")

	client := &http.Client{}
	resp, err := client.Do(req)
	if err != nil {
		return "", fmt.Errorf("request failed: %v", err)
	}
	defer resp.Body.Close()

	body, _ := io.ReadAll(resp.Body)

	if resp.StatusCode != http.StatusCreated {
		return "", fmt.Errorf("%d %s\n  Details: %s", resp.StatusCode, resp.Status, string(body))
	}

	var result JiraResponse
	if err := json.Unmarshal(body, &result); err != nil {
		return "", fmt.Errorf("failed to parse response: %v", err)
	}

	return result.Key, nil
}

func joinStrings(arr []string) string {
	if len(arr) == 0 {
		return ""
	}
	result := arr[0]
	for i := 1; i < len(arr); i++ {
		result += ", " + arr[i]
	}
	return result
}

func strings(char string, count int) string {
	result := ""
	for i := 0; i < count; i++ {
		result += char
	}
	return result
}
